api/include/opentelemetry/context/context.h

```diff
@@ -0,0 +1,205 @@
+#pragma once
+
+#include <atomic>
+
+#include "opentelemetry/context/context_value.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/trace/key_value_iterable_view.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+
+// The context class provides a context identifier. Is built as a linked list
+// of context nodes which each have access to the nodes before it, but not 
+// after. So each time a write is performed a new context is added to the list.
+// Within each context node there is another linked list which holds the key
+// and value data for each node.
+class Context
+{
+
+public:
+  
+  // A linked list to contain the keys and values of this context node 
+  class DataList {
+    private:
+      friend class Context;
+ 
+      DataList *next_;
+      
+      char key_[50];
+      
+      ContextValue value_;
+  };
+  
+  Context(){
+    next_ = NULL;
+    head_ = NULL;
+  }
+
+  // Creates a context object from a map of keys and identifiers, this will
+  // be the head of the context object linked list.
+  template <class T , nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context(const T &keys_and_values)  
+  {
+    head_ = BuildList(keys_and_values);
+    next_ = NULL;
+  }
+  
+  // Creates a context object from a key and value, this will be the head
+  // of the context object linked list.
+  Context(nostd::string_view key, ContextValue value)  
+  {
+    head_ = BuildHead(key, value);
+    next_ = NULL;
+  }
+
+  // Accepts a new iterable and then returns a new context that
+  // contains the new key and value data.
+  template <class T , nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context* SetValues(T &values) noexcept
+  {
+    Context *new_node = new Context(values, this);
+    return new_node;
+  }
+  
+  // Accepts a new iterable and then returns a new context that
+  // contains the new key and value data.
+  Context* SetValue(nostd::string_view key, ContextValue value) noexcept
+  {
+    Context *new_node = new Context(key,value, this);
+    return new_node;
+  }
+
+  // Returns the value associated with the passed in key.
+  common::AttributeValue GetValue(nostd::string_view key)
+  {
+    Context *context_trace = this;
+    
+    //Iterate through the context nodes
+    while(context_trace != NULL){
+      DataList *data_trace = context_trace->head_;
+      
+      //Iterate through the internal data nodes
+      while(data_trace != NULL){
+        if(strncmp(key.data(), data_trace->key_, 50) == 0){
```

Comment:
There's a gotcha with string_view::data() -- because string_view has it's own size, the `const char*` data does not have to be null terminated. I think it's also possible for it to container null characters.

See "notes" on https://en.cppreference.com/w/cpp/string/basic_string_view/data

How about keeping a string_view in DataList nodes that's constructed with [`string_view(const CharT* s, size_type count)`](https://en.cppreference.com/w/cpp/string/basic_string_view/basic_string_view) where `s` is the local copy of the data and `count` is the size of the incoming key?

Then you could do the comparison trivially and generally just work with a string_view (except being careful to copy the data over when constructing the DataList).


 (I'm assuming nostd::string_view is like std::string_view here)

Reply:
I tried to avoid added nostd::string_view as a member variable, just for safety due to the state of OT. And just kept a char * and the length of the string.